api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go

```diff
@@ -179,11 +205,36 @@ func (i *MaterializeStageProvider) Process(ctx context.Context, mgrContext conte
 						mLog.Errorf("Failed to unmarshal solution state for catalog %s: %s: %s", name, err.Error())
 						return outputs, false, err
 					}
-					// If inner solution defines a display name, use it as the name
-					if solutionState.Spec.DisplayName != "" {
-						solutionState.ObjectMeta.Name = solutionState.Spec.DisplayName
+
+					if solutionState.ObjectMeta.Name == "" {
+						mLog.Errorf("Solution name is empty: catalog - %s", name)
+						return outputs, false, v1alpha2.NewCOAError(nil, fmt.Sprintf("Empty solution name: catalog - %s", name), v1alpha2.BadRequest)
 					}
-					solutionState.ObjectMeta = updateObjectMeta(solutionState.ObjectMeta, inputs, name)
+
+					solutionName := solutionState.ObjectMeta.Name
+					parts := strings.Split(solutionName, ":")
+					if len(parts) == 2 {
+						solutionState.Spec.RootResource = parts[0]
+						solutionState.Spec.Version = parts[1]
+					} else {
+						mLog.Errorf("Solution name is invalid: solution - %s, catalog - %s", solutionName, name)
+						return outputs, false, v1alpha2.NewCOAError(nil, fmt.Sprintf("Invalid solution name: catalog - %s", name), v1alpha2.BadRequest)
+					}
+
+					mLog.Debugf("  P (Materialize Processor): check solution contains %v, namespace %s", solutionState.Spec.RootResource, namespace)
+					_, err := i.ApiClient.GetSolutionContainer(ctx, solutionState.Spec.RootResource, namespace, i.Config.User, i.Config.Password)
+					if err != nil && strings.Contains(err.Error(), constants.NotFound) {
```

Comment:
NotFound is an error we can handle. So you can modify the log statement to make it more clear, "failed to get xxxx with error NotFound, try to create ...". How about the error we cannot handle, do you return the error immediately?

Reply:
No. Just to do upsert and webhook will throw out the container doesn't exist. Do you think we need to throw here if not NotFound for pre-handling.

Reply:
As a client, if you are trying to check the parent at first and create if not exists. Why we proceed the child resource creation on any errors other than NotFound? 
This step is ensuring the parent resource's existence. If we get an error during the check, we need to stop the flow and fail as earlier as possible. Or you can add retry to handle the transient error.

Reply:
Updated to throw if not `NotFound`.